examples/tensorflow_hello_world/helloworld.py

```diff
@@ -1,11 +1,211 @@
+
+
+# iterate through each partition (a client's training data)
+# and init a Participant class
+def init_participants(xy_train_partitions, model_provider, xy_validation):
+    potential_participants = []
+    for client_id, xy_train in enumerate(xy_train_partitions):
+        participant = Participant(
+            cid=client_id,
+            model_provider=model_provider,
+            xy_train=xy_train,
+            xy_val=xy_validation,
+            num_classes=NUM_CLASSES,
+            batch_size=B,
+        )
+        potential_participants.append(participant)
+    return potential_participants
+
+
+# init the Coordinator
+def init_coordinator(participants, xy_train_partitions, model_provider, xy_validation):
+
+    # the controller will select the indices of the clients participating in the training
+    # right now it's initialized with all the indices of the clients available
+    num_clients = len(participants)
+    assert num_clients == len(xy_train_partitions)
+    controller = RandomController(num_clients)
+
+    # init the aggregator, Federated Averaging is the state-of-the-art aggregator
+    # see here for more details: https://arxiv.org/pdf/1602.05629.pdf
+    aggregator = FederatedAveragingAgg()
+
+    # the Coordinator will coordinate the training. it will:
+    #   - select a C fraction of potential participants
+    #   - send training jobs to each selected Participant
+    #       who will train its own local data for E local epochs
+    coordinator = Coordinator(
+        controller=controller,
+        model_provider=model_provider,
+        participants=participants,
+        C=C,
+        E=E,
+        xy_val=xy_validation,
+        aggregator=aggregator,
+    )
+    return coordinator
+
+
+# printing some preliminary stats to the terminal
+def print_dataset_stats(xy_train_partitions, xy_validation, xy_test):
+    print(f"\nthere are {len(xy_train_partitions)} client/potential partitions")
+    images_first_client, labels_first_client = xy_train_partitions[0]
+    n_images, height, width = images_first_client.shape
+    print(f"the first client has {n_images} images, of {width}x{height} size")
+    print(f"there are {len(labels_first_client)} labels, one label per image")
+
+    validation_images, _ = xy_validation
+    print(f"---\nthe validation set is made of {len(validation_images)} images")
+
+    test_images, _ = xy_test
+    print(f"the test set is made of {len(test_images)} images\n---")
 
 
-def run():
-    hello = tf.constant("Hello World :)")
-    sess = tf.compat.v1.Session()
-    print(sess.run(hello))
+# printing the validation metrics to the terminal
+def print_validation_metrics(validation_metrics):
+    for round_id in range(R):
+        print(
+            f"validation round: {round_id + 1}: "
+            f"loss: {validation_metrics['val_loss'][round_id]}, "
+            f"accuracy: {validation_metrics['val_acc'][round_id]}"
+        )
 
 
 if __name__ == "__main__":
-    run()
+    app.run(main=main)
```

Comment:
This example doesn't use any `absl-py` features so I'd suggest to remove `absl-py` usage entirely.

Reply:
if i don't use `app.run(main=main)` it doesn't initialise flags and i get the following when i start the script:

`Traceback (most recent call last):
  File "examples/tensorflow_hello_world/helloworld.py", line 248, in <module>
    main()
  File "examples/tensorflow_hello_world/helloworld.py", line 52, in main
    xy_train_partitions, xy_validation, xy_test = load_splits(DATASET_NAME)
  File "/Users/anselmosampietro/Documents/XAIN/xain/xain/datasets/dataset.py", line 62, in load_splits
    dataset_name=dataset_name, get_local_datasets_dir=get_local_datasets_dir
  File "/Users/anselmosampietro/Documents/XAIN/xain/xain/datasets/storage.py", line 125, in load_splits
    local_datasets_dir = get_local_datasets_dir()
  File "/Users/anselmosampietro/Documents/XAIN/xain/xain/datasets/storage.py", line 18, in default_get_local_datasets_dir
    return FLAGS.local_datasets_dir
  File "/Users/anselmosampietro/Documents/XAIN/main37/lib/python3.7/site-packages/absl/flags/_flagvalues.py", line 491, in __getattr__
    raise _exceptions.UnparsedFlagAccessError(error_message)
absl.flags._exceptions.UnparsedFlagAccessError: Trying to access flag --local_datasets_dir before flags were parsed.`